File: Arquivo_Afd_Exemplo_Professor_Para_Adicionar_E_Ler_A_Linguagem/afd.cpp
to#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <stdexcept>

class AFD {
private:
    // Q: conjunto de estados
    std::unordered_set<std::string> estados;

    // Σ: alfabeto
    std::unordered_set<char> alfabeto;

    // δ: função de transição
    std::unordered_map<
        std::string,
        std::unordered_map<char, std::string>
    > transicoes;

    // q₀: estado inicial
    std::string estadoInicial;

    // F: conjunto de estados finais
    std::unordered_set<std::string> estadosFinais;

public:

    void adicionarEstado(const std::string& estado) {
        estados.insert(estado);
    }

    void adicionarSimbolo(char simbolo) {
        alfabeto.insert(simbolo);
    }

    void definirEstadoInicial(const std::string& estado) {
        if (!estados.contains(estado)) {
            throw std::invalid_argument(
                "O estado inicial deve pertencer a Q."
            );
        }

        estadoInicial = estado;
    }

    void adicionarEstadoFinal(const std::string& estado) {
        if (!estados.contains(estado)) {
            throw std::invalid_argument(
                "O estado final deve pertencer a Q."
            );
        }

        estadosFinais.insert(estado);
    }

    void adicionarTransicao(
        const std::string& estado,
        char simbolo,
        const std::string& proximoEstado
    ) {
        if (!estados.contains(estado)) {
            throw std::invalid_argument(
                "Estado de origem inexistente."
            );
        }

        if (!estados.contains(proximoEstado)) {
            throw std::invalid_argument(
                "Estado de destino inexistente."
            );
        }

        if (!alfabeto.contains(simbolo)) {
            throw std::invalid_argument(
                "Simbolo inexistente no alfabeto."
            );
        }

        transicoes[estado][simbolo] = proximoEstado;
    }

    bool aceita(const std::string& palavra) const {

        std::string estado = estadoInicial;

        for (char simbolo : palavra) {

            // Verifica se o símbolo pertence a Σ.
            if (!alfabeto.contains(simbolo)) {
                return false;
            }

            // Procura δ(estado, símbolo).
            auto estadoIt = transicoes.find(estado);

            if (estadoIt == transicoes.end()) {
                return false;
            }

            auto simboloIt = estadoIt->second.find(simbolo);

            if (simboloIt == estadoIt->second.end()) {
                return false;
            }

            // q ← δ(q, símbolo)
            estado = simboloIt->second;
        }

        // w é aceita se q ∈ F.
        return estadosFinais.contains(estado);
    }
};


/*
 * Lê o conteúdo de fonte.lin.
 */
std::string lerArquivo(const std::string& nomeArquivo) {

    std::ifstream arquivo(nomeArquivo);

    if (!arquivo.is_open()) {
        throw std::runtime_error(
            "Nao foi possivel abrir o arquivo: " + nomeArquivo
        );
    }

    // Lê todo o arquivo, inclusive espaços e quebras de linha.
    std::string conteudo(
        (std::istreambuf_iterator<char>(arquivo)),
        std::istreambuf_iterator<char>()
    );

    return conteudo;
}


int main() {

    try {

        // ==================================================
        // CONSTRUÇÃO DO AFD
        // ==================================================

        AFD afd;

        // Q
        afd.adicionarEstado("q0");
        afd.adicionarEstado("q1");
        afd.adicionarEstado("q2");

        // Σ
        afd.adicionarSimbolo('0');
        afd.adicionarSimbolo('1');

        // q₀
        afd.definirEstadoInicial("q0");

        // F
        afd.adicionarEstadoFinal("q2");

        // δ
        afd.adicionarTransicao("q0", '0', "q1");
        afd.adicionarTransicao("q0", '1', "q0");

        afd.adicionarTransicao("q1", '0', "q1");
        afd.adicionarTransicao("q1", '1', "q2");

        afd.adicionarTransicao("q2", '0', "q1");
        afd.adicionarTransicao("q2", '1', "q0");


        // ==================================================
        // LEITURA DO ARQUIVO
        // ==================================================

        const std::string nomeArquivo = "fonte.lin";

        std::string palavra = lerArquivo(nomeArquivo);


        // ==================================================
        // EXECUÇÃO DO AFD
        // ==================================================

        if (afd.aceita(palavra)) {
            std::cout << "ACEITA\n";
        }
        else {
            std::cout << "REJEITADA\n";
        }

    }
    catch (const std::exception& erro) {

        std::cerr << "Erro: "
                  << erro.what()
                  << '\n';

        return 1;
    }

    return 0;
}

File: Afds/Afd_Solis/afd.py
"""
Um AFD é definido formalmente pela quíntupla (Q, Σ, δ, q0, F):

    Q  -> conjunto de estados
    Σ  -> alfabeto (os símbolos que o autômato conhece)
    δ  -> função de transição: dado um estado e um símbolo, diz o próximo estado
    q0 -> estado inicial (onde a leitura começa)
    F  -> conjunto de estados finais (se a leitura terminar num deles, a palavra é aceita)
 
Cada atributo da classe abaixo guarda uma dessas cinco peças.
"""

class AFD:
    def __init__(self):
        #Q : Conjunto de estados.
        self.estados = set()

        #Σ: alfabeto
        self.alfabeto = set()

        # δ: função de transição.
        #É um dicionário de dicionários
        #Transições ["q0"]["5"] = "q1" si
        self.transicoes = {}

        #q0: Estado inicial. Começa vazio até ser definido
        self.estado_inicial = None

        #F: Conjunto de estados finais
        self.estados_finais = set()

    #Métodos para Construção : Montam Q, Σ, δ, q0 e F antes de usar o AFD

    def adicionar_estado(self, estado):
        #Inclui um estado em Q
        #Por enquanto não precisa de validação
        self.estados.add(estado)

    def adicionar_simbolo(self, simbolo):
        # Inclui um símbolo em Σ.
        #Aqui se define quem é o alfabeto
        self.alfabeto.add(simbolo)

    def definir_estado_inicial(self, estado):
        #Aceita como Q0 um estado que já foi declarado em Q
        if estado not in self.estados:
           raise ValueError("O estado inicial dever pertencer a Q")
        self.estado_inicial = estado

    def adicionar_estado_final(self, estado):
        #Aceita em F um estado que foi declarado em Q
        if estado not in self.estados:
            raise ValueError("O estado final deve pertencer a Q")
        self.estados_finais.add(estado)

    def adicionar_transicao(self, estado, simbolo, proximo_estado):
        #Registra uma transição δ(estado, simbolo) = proximo_estado.

        #1) O estado de origem precisa existir em Q
        if estado not in self.estados:
            raise ValueError("Estado de origem inexistente")

        #2) O estado de destino precisa existir em Q
        if proximo_estado not in self.estados:
            raise ValueError("Estado de destino inexistente")

        #3) O símbolo precisa pertencer ao Σ
        if simbolo not in self.alfabeto:
            raise ValueError("Símbolo inexistente no alfabeto")

        #Passou nas três verificações, ele grava a transição

        #setdefault cria o dicionário interno do estado se ele ainda não existir

        self.transicoes.setdefault(estado, {})[simbolo] = proximo_estado

    #Execução - Usa o AFD já montado pra testar uma palavra

    def aceita(self, palavra):
        #A leitura sempre começa no estado inicial (Q0)
        estado = self.estado_inicial

        #Lê a palavra um símbolo por vez, da esquerda para a direita
        
        for simbolo in palavra:

            #Se o símbolo não pertencer ao Σ, a palavra é rejeitada na hora

            if simbolo not in self.alfabeto:
                return False

            #Procura as transições que saem do estado atual

            transicoes_do_estado = self.transicoes.get(estado)

            #Se não tem nehuma transição saindo deste estado, ou nenhuma com esse símbolo, não tem aonde ir, é rejeitado

            if transicoes_do_estado is None or simbolo not in transicoes_do_estado:
                return False

            #Avança : q <- δ(q, símbolo)
            estado = transicoes_do_estado[simbolo]

        #Terminou de ler a palavra inteira e ela é aceita somente se o estado onde paraou pertence a F
        
        return estado in self.estados_finais